### PR TITLE
[ZEPPELIN-3092] GitHub Integration

### PR DESCRIPTION
**Pull request by mohamagdy**
### What is this PR for?
GitHub integration as a storage for notebooks.

### What type of PR is it?
Feature

### What is the Jira issue?
[ZEPPELIN-3092](https://issues.apache.org/jira/browse/ZEPPELIN-3092)

### How should this be tested?
1. Change the configuration in `zeppelin-site.xml` to enable GitHub integration (add GitHub url, username, access token and origin) as described in https://github.com/apache/zeppelin/compare/master...mohamagdy:zeppelin-3092-remote-github-integration?expand=1#diff-89104d48f0358450399a6f679bba9c4f
2. Start the Zeppelin server
3. Open an existing notebook or create a new notebook
4. Do some changes to the notebook, for example add a new paragraph
5. Click on the versioning button on the top menu to commit and save changes
6. Checkout the changes in the GitHub repository. The changes should be reflected 

### Questions:
* **Does the licenses files need update?**
No

* **Is there breaking changes for older versions?**
No

* **Does this needs documentation?**
Yes. Documentation is updated as part of the pull request.

### CODE REVIEW

**Comment by mohamagdy** [comment]
@felixcheung, the `GitHubNotebookRepo` inherits from `GitNotebookRepo` but the difference is, after the notebook is saved (and committed), the `GitHubNotebookRepo` updates (pushes the local commits) a remote GitHub repository (notebooks repository on GitHub) configured in the `zeppelin-site.xml`.

Also, when `GitHubNotebookRepo` is configured as a storage, everytime the notebook is loaded, the latest changes is pulled from the remote GitHub repository.

Does this make sense?

File: docs/setup/storage/storage.md [hunk 1]
```diff
@@ -361,3 +362,44 @@ export ZEPPELIN_NOTEBOOK_MONGO_AUTOIMPORT=true
...
+</property>
+
+<property>
+  <name>zeppelin.notebook.git.remote.access-token</name>
+  <value>1b29e853bc73b90af0d280233101f91b9105d7cb</value>
+  <description>remote Git repository password</description>
+</property>
+
+<property>
+  <name>zeppelin.notebook.git.remote.remote</name>
```
Comment (Leemoonsoo) [inline docs/setup/storage/storage.md]: This supposed to be `zeppelin.notebook.git.remote.origin`?

File: zeppelin-zengine/pom.xml [hunk 2]
```diff
@@ -625,6 +625,11 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+        <groupId>org.eclipse.jgit</groupId>
+        <artifactId>org.eclipse.jgit</artifactId>
+        <version>4.3.1.201605051710-r</version>
```
Comment (Leemoonsoo) [inline zeppelin-zengine/pom.xml]: can we use `${eclipse.jgit.version}` instead?

File: docs/setup/storage/storage.md (hunk 1 above)
Comment (mohamagdy) [reply docs/setup/storage/storage.md]: Good catch 👍 ... Fixed in https://github.com/apache/zeppelin/pull/2700/commits/2c1cf743b54861e9bd53e483f3c873a77dbba1b1

File: zeppelin-zengine/pom.xml (hunk 2 above)
Comment (mohamagdy) [reply zeppelin-zengine/pom.xml]: Unfortunately, version `4.1.1.201511131810-r` which is the value of `${eclipse.jgit.version}` does not contain some classes such as `org.eclipse.jgit.api.RemoteAddCommand`. If possible, I can change the value of `eclipse.jgit.version` to use `4.3.1.201605051710-r` instead of `4.1.1.201511131810-r`. What do you think?

**Comment by mohamagdy** [comment]
@Leemoonsoo, thanks a lot for reviewing and testing the branch.

I will fix the Jenkins job with the information you provided.

Regarding the test files, the tests for the `GitNotebookRepo` was not working because those notebooks were missing from the `src` folder, so I added them. I ran the tests using Intellij.

File: zeppelin-zengine/pom.xml (hunk 2 above)
Comment (jongyoul) [reply zeppelin-zengine/pom.xml]: I think we'd better upgrade the version of `jgit` to 4.3.x

**Comment by jongyoul** [comment]
Can you restart some failed job on https://travis-ci.org/mohamagdy/zeppelin/builds/317797015 ?

File: zeppelin-server/src/test/java/org/apache/zeppelin/AbstractZeppelinIT.java [hunk 3]
```diff
@@ -42,8 +42,8 @@
   protected final static Logger LOG = LoggerFactory.getLogger(AbstractZeppelinIT.class);
   protected static final long MIN_IMPLICIT_WAIT = 5;
   protected static final long MAX_IMPLICIT_WAIT = 30;
-  protected static final long MAX_BROWSER_TIMEOUT_SEC = 30;
-  protected static final long MAX_PARAGRAPH_TIMEOUT_SEC = 120;
+  protected static final long MAX_BROWSER_TIMEOUT_SEC = 180;
```
Comment (jongyoul) [inline zeppelin-server/src/test/java/org/apache/zeppelin/AbstractZeppelinIT.java]: I'm not sure why you change it

Comment (mohamagdy) [reply zeppelin-server/src/test/java/org/apache/zeppelin/AbstractZeppelinIT.java]: I was trying to fix the `timeout` errors that fails on Travis. It didn't work. I will revert those changes back.

File: zeppelin-server/src/main/resources/zeppelin-site.xml [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+
+<property>
+  <name>zeppelin.notebook.homescreen.hide</name>
+  <value>false</value>
+  <description>hide homescreen notebook from list when this value set to true</description>
+</property>
+
+
+<!-- Amazon S3 notebook storage -->
+<!-- Creates the following directory structure: s3://{bucket}/{username}/{notebook-id}/note.json -->
```
Comment (felixcheung) [inline zeppelin-server/src/main/resources/zeppelin-site.xml]: we haven't need this file before, why do we need it now?
also why are lines commented out?

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
+package org.apache.zeppelin.notebook.repo;
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java]: need to add ASF header

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+  private void configureRemoteStream() {
+    try {
+      LOG.debug("Setting up remote stream");
+      RemoteAddCommand remoteAddCommand = git.remoteAdd();
+      remoteAddCommand.setName(zeppelinConfiguration.getZeppelinNotebookGitRemoteOrigin());
+      remoteAddCommand.setUri(new URIish(zeppelinConfiguration.getZeppelinNotebookGitURL()));
+      remoteAddCommand.call();
+    } catch (GitAPIException e) {
+      e.printStackTrace();
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java]: don't printStackTrace, call LOG.error("", e)

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    } catch (URISyntaxException e) {
+      e.printStackTrace();
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java]: ditto

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+          zeppelinConfiguration.getZeppelinNotebookGitUsername(),
+          zeppelinConfiguration.getZeppelinNotebookGitAccessToken()
+        )
+      );
+
+      pullCommand.call();
+
+    } catch (GitAPIException e) {
+      LOG.error("Error when pulling latest changes from remote repository");
+      e.printStackTrace();
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java]: LOG.error("Error when pulling latest changes from remote repository", e);

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        new UsernamePasswordCredentialsProvider(
+          zeppelinConfiguration.getZeppelinNotebookGitUsername(),
+          zeppelinConfiguration.getZeppelinNotebookGitAccessToken()
+        )
+      );
+
+      pushCommand.call();
+    } catch (GitAPIException e) {
+      LOG.error("Error when pushing latest changes from remote repository");
+      e.printStackTrace();
```
Comment (felixcheung) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java]: LOG.error("Error when pushing latest changes from remote repository", e);

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepoTest.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
+package org.apache.zeppelin.notebook.repo;
```
Comment (felixcheung) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepoTest.java]: + ASF header

File: zeppelin-server/src/main/resources/zeppelin-site.xml (hunk 4 above)
Comment (mohamagdy) [reply zeppelin-server/src/main/resources/zeppelin-site.xml]: The file is copied to this path because I had troubles running the `ZeppelinServer` locally. It was missing this file.

Those lines are commented as it is copied from `conf/zeppelin-site.xml.template` https://github.com/apache/zeppelin/blob/master/conf/zeppelin-site.xml.template#L71-L73.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java (hunk 5 above)
Comment (mohamagdy) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java]: 👍 ... Added here https://github.com/apache/zeppelin/pull/2700/commits/62ba96ec0274ec7386fa6531f6bd13191fb09d78

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepoTest.java (hunk 10 above)
Comment (mohamagdy) [reply zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepoTest.java]: 👍 ... Added here https://github.com/apache/zeppelin/pull/2700/commits/62ba96ec0274ec7386fa6531f6bd13191fb09d78

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java (hunk 6 above)
Comment (mohamagdy) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java]: 👍  ... Fixed here https://github.com/apache/zeppelin/pull/2700/commits/9101e58f288cd5617367963b26382a39d9cc26d6

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java (hunk 7 above)
Comment (mohamagdy) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java]: 👍  ... Fixed here https://github.com/apache/zeppelin/pull/2700/commits/9101e58f288cd5617367963b26382a39d9cc26d6

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java (hunk 8 above)
Comment (mohamagdy) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java]: 👍  ... Fixed here https://github.com/apache/zeppelin/pull/2700/commits/9101e58f288cd5617367963b26382a39d9cc26d6

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java (hunk 9 above)
Comment (mohamagdy) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java]: 👍  ... Fixed here https://github.com/apache/zeppelin/pull/2700/commits/9101e58f288cd5617367963b26382a39d9cc26d6

File: zeppelin-server/src/main/resources/zeppelin-site.xml (hunk 4 above)
Comment (felixcheung) [reply zeppelin-server/src/main/resources/zeppelin-site.xml]: I'm wondering how other would think about this - I think we don't include default xml config (and only .xml.template instead) for a reason

**Comment by felixcheung** [comment]
and we should probably check why RAT doesn't pick up the missing ASF header for new files?

File: zeppelin-server/src/main/resources/zeppelin-site.xml (hunk 4 above)
Comment (mohamagdy) [reply zeppelin-server/src/main/resources/zeppelin-site.xml]: I understand your point. I would suggest removing this file and adda step to the README to copy this file and do the proper configuration changes locally when running `ZeppelinServer` in development mode.

Comment (felixcheung) [reply zeppelin-server/src/main/resources/zeppelin-site.xml]: ok

Comment (mohamagdy) [reply zeppelin-server/src/main/resources/zeppelin-site.xml]: 👍 

1. Removed the `zeppelin-site.xml` from Zeppelin Server resources
2. Added instructions to the `how_to_contribute.md` to copy the template file from `conf` directory to resources directory
3. Added the `zeppelin-site.xml` in the resources directory to `.gitignore`

https://github.com/apache/zeppelin/pull/2700/commits/e98d1b0f2c1bb51c261807365f5c3d0d31a7d46c and https://github.com/apache/zeppelin/pull/2700/commits/b4fe44d7453dbb734910f06a04d45a5657364ae7.

**Comment by felixcheung** [comment]
looks like Jenkins has lost the test run result - could you kick that off again? you can do that by closing and reopening this PR.

**Comment by felixcheung** [comment]
@mohamagdy could you check - there are a bunch of errors https://travis-ci.org/mohamagdy/zeppelin/builds/324495310

File: .gitignore [hunk 11]
```diff
@@ -18,6 +18,9 @@ spark-1.*-bin-hadoop*
 
 lens/lens-cli-hist.log
 
+# Zeppelin server
+zeppelin-server/local-repo
+zeppelin-server/src/main/resources/zeppelin-site.xml
```
Comment (zjffdu) [inline .gitignore]: why this is ignored ?

Comment (mohamagdy) [reply .gitignore]: There was a discussion about this file here https://github.com/apache/zeppelin/pull/2700#discussion_r158531529.

File: docs/development/contribution/how_to_contribute_code.md [hunk 12]
```diff
@@ -104,6 +110,12 @@ bin/zeppelin-daemon start
 
 Server will be run on [http://localhost:8080](http://localhost:8080).
 
+#### Option 3 - IDE
+
+1. Copy the `conf/zeppelin-site.xml.template` to `zeppelin-server/src/main/resources/zeppelin-site.xml` and change the configurations in this file if required
+2. `ZeppelinServer.java` Main class
+
+
```
Comment (zjffdu) [inline docs/development/contribution/how_to_contribute_code.md]: Thanks for the contribution for how to debugging zeppelin-server. Personally I would recommend to use Daemon script, because daemon script would also do some other set up for zeppelin server (like run zeppelin-env.sh), and other approaches (Command Line & IDE) don't do that.

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java [hunk 13]
```diff
@@ -824,8 +824,8 @@ private void sendNote(NotebookSocket conn, HashSet<String> userAndRoles, Noteboo
     String user = fromMessage.principal;
 
     Note note = notebook.getNote(noteId);
-    if (note != null) {
 
+    if (note != null) {
```
Comment (zjffdu) [inline zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: unnecessary change ?

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+  private void updateRemoteStream() {
+    LOG.debug("Updating remote stream");
+
+    pullFromRemoteStream();
+    pushToRemoteSteam();
+  }
+
+  private void pullFromRemoteStream() {
+    try {
+      LOG.debug("Pull latest changed from remote stream");
```
Comment (zjffdu) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitHubNotebookRepo.java]: What happens when it fails to pull from remote repo due to conflict, it is very likely to happen when note is shared between users ? Does the UI display any error message ?

File: zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java (hunk 13 above)
Comment (mohamagdy) [reply zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java]: Yes, nothing changed here

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java [hunk 15]
```diff
@@ -34,6 +34,7 @@
 import com.google.common.base.Predicate;
 import com.google.common.collect.FluentIterable;
 import com.google.common.collect.Sets;
+import org.apache.commons.lang.StringUtils;
```
[the remark this answers is not in the document]
Comment (mohamagdy) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java]: Good catch. Fixed in https://github.com/apache/zeppelin/pull/2700/commits/b445960e8ef96686f56f016a02083895b4c8d1d7. Optimized all the imports for the `Notebook` class

File: docs/development/contribution/how_to_contribute_code.md (hunk 12 above)
Comment (mohamagdy) [reply docs/development/contribution/how_to_contribute_code.md]: In my case, it was enough to run the `ZeppelinServer.java` Main class to test the changes on the fly. It required much less effort from my side to set up the IDE to run the Main class. Though it worth adding this information to the `how_to_contribute` file.

**Comment by zjffdu** [comment]
ping @mohamagdy

**Comment by mohamagdy** [comment]
@zjffdu all good. I closed and reopened the pull request.

I would suggest making the process of rerunning Jenkins easier than closing and reopening the pull request. This also triggers Travis tests which take around 1.5 hours. Maybe adding a `build` button in Jenkins so that one can rerun the job? What do you think?

**Comment by zjffdu** [comment]
It is awsome if there's such button. But I don't know how to make such button. BTW usually I would force push a dummy commit to trigger the build. And I believe we should also fix these flaky test which cause inconvenience for developers.

**Comment by nighthawk6389** [comment]
Would love to see this merged!
